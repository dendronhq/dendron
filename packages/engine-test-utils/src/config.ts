import { DConfig, readYAML, writeYAML } from "@dendronhq/common-server";
import { IntermediateDendronConfig } from "@dendronhq/common-all";

export class TestConfigUtils {
  static getConfig = (opts: { wsRoot: string }) => {
    const configPath = DConfig.configPath(opts.wsRoot);
    const config = readYAML(configPath) as IntermediateDendronConfig;
    return config;
  };

  static withConfig = (
    func: (config: IntermediateDendronConfig) => IntermediateDendronConfig,
    opts: { wsRoot: string }
  ) => {
    const config = TestConfigUtils.getConfig(opts);

    const newConfig = func(config);
    TestConfigUtils.writeConfig({ config: newConfig, wsRoot: opts.wsRoot });
    return newConfig;
  };

  static writeConfig = (opts: {
    config: IntermediateDendronConfig;
    wsRoot: string;
  }) => {
    const configPath = DConfig.configPath(opts.wsRoot);
    return writeYAML(configPath, opts.config);
  };
}
