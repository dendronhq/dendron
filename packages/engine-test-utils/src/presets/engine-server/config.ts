import { CONSTANTS, ConfigUtils } from "@dendronhq/common-all";
import { DConfig, readYAML, writeYAML } from "@dendronhq/common-server";
import { TestPresetEntryV4 } from "@dendronhq/common-test-utils";
import path from "path";

function genDefaultConfig() {
  return ConfigUtils.genDefaultConfig();
}

const WRITE = {
  NEW_CONFIG: new TestPresetEntryV4(async ({ wsRoot }) => {
    const config = genDefaultConfig();
    ConfigUtils.setPublishProp(config, "copyAssets", false);
    await DConfig.writeConfig({ wsRoot, config });
    const cpath = path.join(wsRoot, CONSTANTS.DENDRON_CONFIG_FILE);
    const configOnFile = readYAML(cpath);
    return [
      {
        actual: configOnFile,
        expected: config,
      },
    ];
  }),
};

const GET = {
  DEFAULT_CONFIG: new TestPresetEntryV4(async ({ wsRoot }) => {
    const cpath = path.join(wsRoot, CONSTANTS.DENDRON_CONFIG_FILE);
    writeYAML(cpath, genDefaultConfig());
    const config = DConfig.readConfigSync(wsRoot);
    return [
      {
        actual: config,
        expected: genDefaultConfig(),
      },
    ];
  }),
};

export const ENGINE_CONFIG_PRESETS = {
  WRITE,
  GET,
};
