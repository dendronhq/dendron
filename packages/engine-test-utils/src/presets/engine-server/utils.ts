import { NoteProps, SchemaUtils } from "@dendronhq/common-all";
import {
  CreateNoteFactory,
  NOTE_PRESETS_V4,
  NoteTestUtilsV4,
  PreSetupHookFunction,
  SCHEMA_PRESETS_V4,
} from "@dendronhq/common-test-utils";
import _ from "lodash";
import { resolvePath } from "@dendronhq/common-server";
import path from "path";
import * as fs from "fs";

/**
 * Notes created:
 *   - vault1:
 *     - foo
 *     - foo.ch1
 *   - vault2:
 *     - bar
 */
export const setupBasic: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
  extra,
}) => {
  const vault = vaults[0];
  // TODO: HACK
  let props;
  if (extra?.idv2) {
    props = { id: "foo-id" };
  }
  await NOTE_PRESETS_V4.NOTE_SIMPLE.create({
    vault,
    wsRoot,
    props,
  });
  await NOTE_PRESETS_V4.NOTE_SIMPLE_CHILD.create({
    vault,
    wsRoot,
  });
  await NOTE_PRESETS_V4.NOTE_SIMPLE_OTHER.create({
    vault,
    wsRoot,
  });
  await SCHEMA_PRESETS_V4.SCHEMA_SIMPLE.create({ vault, wsRoot });
};

/**
 <pre>
 /vault1
 ├── bar.ch1.gch1.ggch1.md
 ├── bar.ch1.gch1.md
 ├── bar.ch1.md
 ├── bar.md
 ├── foo.ch1.gch1.ggch1.md
 ├── foo.ch1.gch1.md
 ├── foo.ch1.gch2.md
 ├── foo.ch1.md
 ├── foo.ch2.md
 ├── foo.md
 ├── root.md
 └── root.schema.yml

/vault2
 ├── root.md
 └── root.schema.yml

 /vault3
 ├── root.md
 └── root.schema.yml
</pre>
 * */
export const setupHierarchyForLookupTests: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
}) => {
  const opts = {
    vault: vaults[0],
    wsRoot,
  };
  const fnames = [
    "foo",
    "foo.ch1",
    "foo.ch1.gch1",
    "foo.ch1.gch1.ggch1",
    "foo.ch1.gch2",
    "foo.ch2",
    "bar",
    "bar.ch1",
    "bar.ch1.gch1",
    "bar.ch1.gch1.ggch1",
    "goo.ends-with-ch1.no-ch1-by-itself",
  ];

  for (const fname of fnames) {
    await CreateNoteFactory({ fname, body: `${fname} body` }).create(opts);
  }
};

export const setupJournals: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
}) => {
  const vault = vaults[0];
  const names = [
    "daily",
    "daily.journal",
    "daily.journal.2020",
    "daily.journal.2020.07",
    "daily.journal.2020.07.01.one",
    "daily.journal.2020.07.05.two",
  ];
  return Promise.all(
    names.map((fname) => {
      return NoteTestUtilsV4.createNote({
        wsRoot,
        vault,
        fname,
      });
    })
  );
};

// Workspace will look like:
// .
// ├── dendron.code-workspace
// ├── dendron.yml
// ├── vault1
// │   ├── foo.ch1.md
// │   ├── foo.md
// │   ├── foo.schema.yml
// │   ├── root.md
// │   └── root.schema.yml
// ├── vault2
// │   ├── bar.md
// │   ├── root.md
// │   └── root.schema.yml
// └── vault3
//     ├── root.md
//     └── root.schema.yml
export const setupBasicMulti: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
}) => {
  const vault1 = _.find(vaults, { fsPath: "vault1" })!;
  const vault2 = _.find(vaults, { fsPath: "vault2" })!;
  await NOTE_PRESETS_V4.NOTE_SIMPLE.create({
    vault: vault1,
    wsRoot,
  });
  await NOTE_PRESETS_V4.NOTE_SIMPLE_CHILD.create({
    vault: vault1,
    wsRoot,
  });
  await NOTE_PRESETS_V4.NOTE_SIMPLE_OTHER.create({
    vault: vault2,
    wsRoot,
  });
  await SCHEMA_PRESETS_V4.SCHEMA_SIMPLE.create({ vault: vault1, wsRoot });
};

/**
 *
 * - foo
 * ```
 * ![[foo.one]]
 * ```
 *
 * - foo.one
 * ```
 * # Foo.One
 * ![[foo.two]]
 * Regular wikilink: [[foo.two]]
 * ```
 *
 * - foo.two
 * ```
 * # Foo.Two
 * blah
 * ```
 *
 *
 */
export const setupNoteRefRecursive: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
  extra,
}) => {
  const vault = vaults[0];
  // TODO: HACK
  let props;
  if (extra?.idv2) {
    props = { id: "foo-id" };
  }
  await NOTE_PRESETS_V4.NOTE_SIMPLE.create({
    vault,
    wsRoot,
    body: "![[foo.one]]",
    props,
  });
  await NoteTestUtilsV4.createNote({
    vault,
    fname: "foo.one",
    body: ["# Foo.One", `![[foo.two]]`, `Regular wikilink: [[foo.two]]`].join(
      "\n"
    ),
    wsRoot,
    props: {
      id: "foo.one-id",
    },
  });
  await NoteTestUtilsV4.createNote({
    vault,
    fname: "foo.two",
    body: ["# Foo.Two", `blah`].join("\n"),
    wsRoot,
  });
};

/**
 * Diamond schema is laid out such that:
 *    bar
 *   /   \
 * ch1   ch2
 *   \   /
 *    gch
 * */
export const setupSchemaWithDiamondGrandchildren: PreSetupHookFunction = async (
  opts
) => {
  await setupBasic(opts);
  const { wsRoot, vaults } = opts;
  const vault = vaults[0];
  NoteTestUtilsV4.createSchema({
    fname: "bar",
    wsRoot,
    vault,
    modifier: (schema) => {
      const schemas = [
        SchemaUtils.createFromSchemaOpts({
          id: "bar",
          parent: "root",
          fname: "bar",
          children: ["ch1", "ch2"],
          vault,
        }),
        SchemaUtils.createFromSchemaRaw({
          id: "ch1",
          children: ["gch"],
          vault,
        }),
        SchemaUtils.createFromSchemaRaw({
          id: "ch2",
          children: ["gch"],
          vault,
        }),
        SchemaUtils.createFromSchemaRaw({
          id: "gch",
          template: { id: "template.gch", type: "note" },
          vault,
        }),
      ];
      schemas.map((s) => {
        schema.schemas[s.id] = s;
      });
      return schema;
    },
  });
  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "gch template",
    fname: "template.gch",
    vault,
  });
};

export const setupSchemaPreseet: PreSetupHookFunction = async (opts) => {
  await setupBasic(opts);
  const { wsRoot, vaults } = opts;
  const vault = vaults[0];
  NoteTestUtilsV4.createSchema({
    fname: "bar",
    wsRoot,
    vault,
    modifier: (schema) => {
      const schemas = [
        SchemaUtils.createFromSchemaOpts({
          id: "bar",
          parent: "root",
          fname: "bar",
          children: ["ch1", "ch2"],
          vault,
        }),
        SchemaUtils.createFromSchemaRaw({
          id: "ch1",
          template: { id: "bar.template.ch1", type: "note" },
          vault,
        }),
        SchemaUtils.createFromSchemaRaw({
          id: "ch2",
          template: { id: "bar.template.ch2", type: "note" },
          namespace: true,
          vault,
        }),
      ];
      schemas.map((s) => {
        schema.schemas[s.id] = s;
      });
      return schema;
    },
  });
  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "ch1 template",
    fname: "bar.template.ch1",
    props: { tags: "tag-foo" },
    vault,
  });
  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "ch2 template",
    fname: "bar.template.ch2",
    vault,
  });
};

/**
 * Diamond schema is laid out such that:
 *    bar
 *   /   \
 * ch1   ch2 (namespace: true)
 *   \   /
 *    gch
 * */
export const setupSchemaWithDiamondAndParentNamespace: PreSetupHookFunction =
  async (opts) => {
    await setupBasic(opts);
    const { wsRoot, vaults } = opts;
    const vault1 = vaults[0];

    const withDiamond = path.join(
      resolvePath(vault1.fsPath, wsRoot),
      "withDiamond.schema.yml"
    );
    fs.writeFileSync(
      withDiamond,
      `
version: 1
schemas:
  - id: withDiamond
    children:
      - ch1
      - ch2
    title: withDiamond
    parent: root
  - id: ch1
    children:
      - gch
  - id: ch2
    namespace: true
    children:
      - gch
  - id: gch
    template: template.test
`
    );

    await NoteTestUtilsV4.createNote({
      wsRoot,
      body: "Template text",
      fname: "template.test",
      vault: vault1,
    });
  };

/**
 * Sets up schema which includes a schema that has Diamond grandchildren
 *
 * */
export const setupSchemaWithIncludeOfDiamond: PreSetupHookFunction = async (
  opts
) => {
  await setupBasic(opts);
  const { wsRoot, vaults } = opts;
  const vault1 = vaults[0];

  const withDiamond = path.join(
    resolvePath(vault1.fsPath, wsRoot),
    "withDiamond.schema.yml"
  );
  fs.writeFileSync(
    withDiamond,
    `
version: 1
schemas:
  - id: withDiamond
    children:
      - ch1
      - ch2
    title: withDiamond
    parent: root
  - id: ch1
    children:
      - gch
  - id: ch2
    children:
      - gch
  - id: gch
    template: template.test
`
  );

  const includesDiamondPath = path.join(
    resolvePath(vault1.fsPath, wsRoot),
    "includesDiamond.schema.yml"
  );
  fs.writeFileSync(
    includesDiamondPath,
    `
version: 1

imports:
  - withDiamond

schemas:
  - id: includesDiamond
    parent: root
    namespace: true
    children:
      - a-ch1
      - a-ch2
  - id: a-ch1
    children:
      - withDiamond.gch
  - id: a-ch2
    children:
      - withDiamond.gch
`
  );

  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "Template text",
    fname: "template.test",
    vault: vault1,
  });
};

/**
 * Sets up workspace which has a schema that uses YAML expansion syntax ('<<' type of expansion). */
export const setupSchemaWithExpansion: PreSetupHookFunction = async (opts) => {
  await setupBasic(opts);
  const { wsRoot, vaults } = opts;
  const vault1 = vaults[0];

  const withExpansion = path.join(
    resolvePath(vault1.fsPath, wsRoot),
    "withExpansion.schema.yml"
  );
  fs.writeFileSync(
    withExpansion,
    `
_anchors:
  projects: &projects
    title: projects
    parent: root
    template: templates.projects
    
version: 1
imports: []

schemas:
  - <<: *projects
    id: proj
  - <<: *projects
    id: op
`
  );

  const inlineSchemaPath = path.join(
    resolvePath(vault1.fsPath, wsRoot),
    "includesExpansion.schema.yml"
  );
  fs.writeFileSync(
    inlineSchemaPath,
    `
version: 1

imports:
  - withExpansion

schemas:
  - id: includer
    parent: root
    namespace: true
    children:
      - withExpansion.proj
`
  );

  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "Template text",
    fname: "templates.projects",
    vault: vault1,
  });
};

export const setupInlineSchema: PreSetupHookFunction = async (opts) => {
  await setupBasic(opts);
  const { wsRoot, vaults } = opts;
  const vault1 = vaults[0];

  const inlineSchemaPath = path.join(
    resolvePath(vault1.fsPath, wsRoot),
    "inlined.schema.yml"
  );
  fs.writeFileSync(
    inlineSchemaPath,
    `
version: 1
imports: 
  - foo
schemas:
  - id: plain_schema
    parent: root
    children:
      - plain_schema_child
      - daily
  - id: daily
    parent: root
    children:
      - id: journal
        children:
          - pattern: "[0-2][0-9][0-9][0-9]"
            title: year
            id: year_id
            children:
              - pattern: "[0-1][0-9]"
                children:
                  - pattern: "[0-3][0-9]"
                    title: day
                    template:
                      id: templates.day
                      type: note
  - id: id_with_imported_child
    children:
      - foo.foo
  - id: with_child_that_has_untyped_template
    children:
      - pattern: has_untyped_template
        template: templates.untyped
  - id: plain_schema_child
    template: templates.example
`
  );

  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "Template text",
    fname: "templates.day",
    vault: vault1,
  });

  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "Template text",
    fname: "templates.example",
    vault: vault1,
  });

  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "Untyped template",
    fname: "templates.untyped",
    vault: vault1,
  });
};

export const setupSchemaPresetWithNamespaceTemplateBase: PreSetupHookFunction =
  async (opts) => {
    await setupBasic(opts);
    const { wsRoot, vaults } = opts;
    const vault1 = vaults[0];
    const vault2 = vaults[1];
    NoteTestUtilsV4.createSchema({
      fname: "journal",
      wsRoot,
      vault: vault1,
      modifier: (schema) => {
        const schemas = [
          SchemaUtils.createFromSchemaOpts({
            id: "daily",
            parent: "root",
            fname: "daily",
            children: ["journal"],
            vault: vault1,
          }),
          SchemaUtils.createFromSchemaRaw({
            id: "journal",
            children: ["year"],
            vault: vault1,
          }),
          SchemaUtils.createFromSchemaRaw({
            id: "year",
            pattern: "[0-2][0-9][0-9][0-9]",
            children: ["month"],
            vault: vault1,
          }),
          SchemaUtils.createFromSchemaRaw({
            id: "month",
            pattern: "[0-9][0-9]",
            children: ["day"],
            vault: vault1,
          }),
          SchemaUtils.createFromSchemaRaw({
            id: "day",
            pattern: "[0-9][0-9]",
            namespace: true,
            template: {
              id: "journal.template",
              type: "note",
            },
            vault: vault2,
          }),
        ];
        schemas.map((s) => {
          schema.schemas[s.id] = s;
        });
        return schema;
      },
    });
    await NoteTestUtilsV4.createNote({
      wsRoot,
      body: "Journal",
      fname: "daily",
      vault: vault2,
    });
    await NoteTestUtilsV4.createNote({
      wsRoot,
      body: "Template text",
      fname: "journal.template",
      vault: vault2,
    });
  };

export const setupSchemaPresetWithNamespaceTemplateMulti: PreSetupHookFunction =
  async (opts) => {
    return setupSchemaPresetWithNamespaceTemplateBase(opts);
  };

export const setupSchemaPresetWithNamespaceTemplate: PreSetupHookFunction =
  async (opts) => {
    const vault = opts.vaults[0];
    return setupSchemaPresetWithNamespaceTemplateBase({
      ...opts,
      vaults: [vault, vault],
    });
  };

export const setupEmpty: PreSetupHookFunction = async ({ vaults, wsRoot }) => {
  const vault = vaults[0];
  await NOTE_PRESETS_V4.NOTE_EMPTY.create({
    vault,
    wsRoot,
  });
};

export const setupLinks: PreSetupHookFunction = async ({ vaults, wsRoot }) => {
  return setupLinksBase({ wsRoot, vaults: [vaults[0], vaults[0]] });
};

export const setupLinksMulti: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
}) => {
  return setupLinksBase({ wsRoot, vaults });
};

export const setupLinksBase: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
}) => {
  const vault1 = vaults[0];
  const vault2 = vaults[1];
  // create note with wikilink
  await NOTE_PRESETS_V4.NOTE_WITH_TARGET.create({
    vault: vault1,
    wsRoot,
  });
  // create note with relative wikilink
  await NOTE_PRESETS_V4.NOTE_WITH_ANCHOR_LINK.create({
    vault: vault2,
    wsRoot,
  });
  // create note with labeld wikilink
  await NoteTestUtilsV4.createNote({
    wsRoot,
    body: "[[some label|beta]]",
    fname: "omega",
    vault: vault1,
  });
};

/** Creates 2 notes with same fname in 2 different vaults, and a note named
 * "test" in second vault with both valid and invalid wikilinks.
 *
 * See [[scratch.2021.07.15.205433.inconsistent-ref-and-link-behavior]] for the
 * invalid behaviors that this is intended to test for. The only difference is
 * that vaultThree is the real vault and vault3 is the bad one.
 *
 * @returns the test note with the wikilinks.
 */
export const setupMultiVaultSameFname: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
}): Promise<NoteProps> => {
  await NoteTestUtilsV4.createNote({
    fname: "eggs",
    vault: vaults[0],
    body: "vault 0",
    wsRoot,
    props: { id: "eggs-vault-0" },
  });
  await NoteTestUtilsV4.createNote({
    fname: "eggs",
    vault: vaults[1],
    body: "vault 1",
    wsRoot,
    props: { id: "eggs-vault-1" },
  });
  return NoteTestUtilsV4.createNote({
    fname: "test",
    vault: vaults[1],
    body: [
      "[[eggs]]", // 7
      "[[dendron://vault1/eggs]]", // 8
      "[[dendron://vault2/eggs]]", // 9
      "[[dendron://vaultThree/eggs]]", // 10
      "[[dendron://vault3/eggs]]", // 11
      "",
      "the test note",
    ].join("\n"),
    wsRoot,
  });
};

export const setupLinksWithVaultBase: PreSetupHookFunction = async ({
  vaults,
  wsRoot,
}) => {
  await setupLinksBase({ vaults, wsRoot });
};

export const setupRefs: PreSetupHookFunction = async ({ vaults, wsRoot }) => {
  const vault = vaults[0];
  // create note with note reference
  await NOTE_PRESETS_V4.NOTE_WITH_NOTE_REF_SIMPLE.create({
    vault,
    wsRoot,
  });
  await NOTE_PRESETS_V4.NOTE_WITH_NOTE_REF_SIMPLE_TARGET.create({
    vault,
    wsRoot,
  });
  // create note with block reference
  await NOTE_PRESETS_V4.NOTE_WITH_BLOCK_REF_SIMPLE.create({
    vault,
    wsRoot,
  });
  // create note with block range reference
  await NOTE_PRESETS_V4.NOTE_WITH_BLOCK_RANGE_REF_SIMPLE.create({
    vault,
    wsRoot,
  });
  // create note with reference offset
  await NOTE_PRESETS_V4.NOTE_WITH_REF_OFFSET.create({
    vault,
    wsRoot,
  });
  // create note with wildcard child reference
  await NOTE_PRESETS_V4.NOTE_WITH_WILDCARD_CHILD_REF.create({
    vault,
    wsRoot,
  });
  // create note with wildcard header reference
  await NOTE_PRESETS_V4.NOTE_WITH_WILDCARD_HEADER_REF.create({
    vault,
    wsRoot,
  });
  // create note with complex wildcard reference
  await NOTE_PRESETS_V4.NOTE_WITH_WILDCARD_COMPLEX.create({
    vault,
    wsRoot,
  });
  // create note with fm variables
  await NOTE_PRESETS_V4.NOTE_WITH_FM_VARIABLES.create({
    vault,
    wsRoot,
  });
};

export const ENGINE_HOOKS_BASE = {
  WITH_LINKS: setupLinksBase,
};

export const ENGINE_HOOKS = {
  setupBasic,
  setupHierarchyForLookupTests,
  setupSchemaPreseet,
  setupSchemaWithDiamondGrandchildren,
  setupSchemaWithIncludeOfDiamond,
  setupSchemaWithDiamondAndParentNamespace,
  setupSchemaPresetWithNamespaceTemplate,
  setupInlineSchema,
  setupSchemaWithExpansion,
  setupNoteRefRecursive,
  setupJournals,
  setupEmpty,
  setupLinks,
  setupRefs,
};

export const ENGINE_HOOKS_MULTI = {
  setupBasicMulti,
  setupLinksMulti,
  setupSchemaPresetWithNamespaceTemplateMulti,
  setupMultiVaultSameFname,
};
