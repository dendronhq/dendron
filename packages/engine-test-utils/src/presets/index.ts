import { DVault, VaultUtils } from "@dendronhq/common-all";
import { NoteTestUtilsV4 } from "@dendronhq/common-test-utils";
import { WorkspaceOpts } from "../engine";

export * from "./engine-server";
export * from "./pods-core";
export * from "./vaults";

export enum SETUP_HOOK_KEYS {
  /**
   * alpha: link(beta)
   * beta: link(alpha)
   */
  WITH_LINKS = "WITH_LINKS",
}

const createLink = (fname: string, opts: { vaultPrefix?: DVault }) => {
  let cVaultPrefix = "";
  if (opts.vaultPrefix) {
    cVaultPrefix = VaultUtils.toURIPrefix(opts.vaultPrefix) + "/";
  }
  return `[[${cVaultPrefix}${fname}]]`;
};

export async function callSetupHook(
  key: SETUP_HOOK_KEYS,
  opts: {
    workspaceType: "single" | "multi";
    withVaultPrefix?: boolean;
  } & WorkspaceOpts
) {
  const { workspaceType, vaults, wsRoot, withVaultPrefix } = opts;
  const isMultiVault = workspaceType !== "single";
  const cVaults = isMultiVault ? vaults : [vaults[0], vaults[0]];

  // WITH LINKS
  if (key === SETUP_HOOK_KEYS.WITH_LINKS) {
    const link1 = createLink("beta", {
      vaultPrefix: withVaultPrefix ? cVaults[1] : undefined,
    });
    await NoteTestUtilsV4.createNote({
      fname: "alpha",
      wsRoot,
      vault: cVaults[0],
      body: link1,
    });
    const link2 = createLink("alpha", {
      vaultPrefix: withVaultPrefix ? cVaults[0] : undefined,
    });
    await NoteTestUtilsV4.createNote({
      fname: "beta",
      wsRoot,
      vault: cVaults[1],
      body: link2,
    });
    if (isMultiVault) {
      await NoteTestUtilsV4.createNote({
        fname: "beta",
        wsRoot,
        vault: cVaults[0],
        body: link2,
        genRandomId: true,
      });
    }
  } else {
    throw Error("not supported key");
  }
}
