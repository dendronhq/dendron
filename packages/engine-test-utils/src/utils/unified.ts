import { checkString } from ".";

export class TestUnifiedUtils {
  static verifyPrivateLink = ({
    contents,
    value,
  }: {
    contents: string;
    value: string;
  }) => {
    return checkString(contents, "color: brown", value + " (Private)");
  };
}
