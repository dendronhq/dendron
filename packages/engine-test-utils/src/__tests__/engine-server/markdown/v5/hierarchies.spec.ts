import { NoteUtils, WorkspaceOpts } from "@dendronhq/common-all";
import { NoteTestUtilsV4 } from "@dendronhq/common-test-utils";
import { DendronASTDest, ProcFlavor } from "@dendronhq/unified";
import { TestConfigUtils } from "../../../../config";
import { ENGINE_HOOKS } from "../../../../presets";
import { checkNotInVFile, checkVFile, createProcCompileTests } from "../utils";
import { getOpts, runTestCases } from "./utils";

describe("WHEN enableChildLinks", () => {
  runTestCases(
    createProcCompileTests({
      name: "enableChildLinks",
      setup: async (opts) => {
        const { proc } = getOpts(opts);
        const txt = `foo test`;
        const resp = await proc.process(txt);
        return { resp, proc };
      },
      verify: {
        [DendronASTDest.HTML]: {
          [ProcFlavor.PUBLISHING]: async ({ extra }) => {
            const { resp } = extra;
            await checkVFile(resp, "Children", "foo.ch1");
          },
        },
      },
      preSetupHook: ENGINE_HOOKS.setupBasic,
    })
  );
});

describe("WHEN note enableChildLinks = false", () => {
  runTestCases(
    createProcCompileTests({
      name: "noEnableChildLinks",
      setup: async (opts) => {
        const { proc } = getOpts(opts);
        const txt = `foo test`;
        const resp = await proc.process(txt);
        return { resp, proc };
      },
      verify: {
        [DendronASTDest.HTML]: {
          [ProcFlavor.PUBLISHING]: async ({ extra }) => {
            const { resp } = extra;
            await checkNotInVFile(resp, "Children", "foo.ch1");
          },
        },
      },
      preSetupHook: async (opts) => {
        await ENGINE_HOOKS.setupBasic(opts);
        await NoteTestUtilsV4.modifyNoteByPath(
          { ...opts, vault: opts.vaults[0], fname: "foo" },
          (note) => {
            NoteUtils.updateNoteLocalConfig(note, "global", {
              enableChildLinks: false,
            });
            return note;
          }
        );
      },
    })
  );
});

describe("WHEN enableHierarchyDisplay is set to true", () => {
  const BASIC_TEXT = "[Ch1](foo.ch1.html)";
  runTestCases(
    createProcCompileTests({
      name: "enableHierarchyDisplay",
      setup: async (opts) => {
        const { proc } = getOpts(opts);

        const resp = await proc.process(BASIC_TEXT);
        return { resp, proc };
      },
      verify: {
        [DendronASTDest.HTML]: {
          [ProcFlavor.PUBLISHING]: async ({ extra }) => {
            const { resp } = extra;
            await checkVFile(resp, "Children");
          },
        },
      },
      preSetupHook: ENGINE_HOOKS.setupBasic,
    })
  );
});

describe("WHEN enableHierarchyDisplay is set to false", () => {
  const preSetupHook = async (opts: WorkspaceOpts) => {
    await ENGINE_HOOKS.setupBasic(opts);
    TestConfigUtils.withConfig((config) => {
      config.publishing!.enableHierarchyDisplay = false;
      return config;
    }, opts);
  };
  const BASIC_TEXT = "[Ch1](foo.ch1.html)";
  runTestCases(
    createProcCompileTests({
      name: "enableHierarchyDisplay",
      setup: async (opts) => {
        const { proc } = getOpts(opts);

        const resp = await proc.process(BASIC_TEXT);
        return { resp, proc };
      },
      verify: {
        [DendronASTDest.HTML]: {
          [ProcFlavor.PUBLISHING]: async ({ extra }) => {
            const { resp } = extra;
            await checkNotInVFile(resp, "Children");
          },
        },
      },
      preSetupHook,
    })
  );
});

describe("WHEN hierarchyDisplayTitle is set to 'Better Children' ", () => {
  const preSetupHook = async (opts: WorkspaceOpts) => {
    await ENGINE_HOOKS.setupBasic(opts);
    TestConfigUtils.withConfig((config) => {
      config.publishing!.hierarchyDisplayTitle = "Better Children";
      return config;
    }, opts);
  };
  const BASIC_TEXT = "[Ch1](foo.ch1.html)";
  runTestCases(
    createProcCompileTests({
      name: "hierarchyDisplayTitle",
      setup: async (opts) => {
        const { proc } = getOpts(opts);

        const resp = await proc.process(BASIC_TEXT);
        return { resp, proc };
      },
      verify: {
        [DendronASTDest.HTML]: {
          [ProcFlavor.PUBLISHING]: async ({ extra }) => {
            const { resp } = extra;
            await checkVFile(resp, "Better Children");
          },
        },
      },
      preSetupHook,
    })
  );
});
