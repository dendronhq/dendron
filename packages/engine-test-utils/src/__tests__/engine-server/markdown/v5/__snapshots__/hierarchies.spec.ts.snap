// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WHEN enableChildLinks "HTML: enableChildLinks: PUBLISHING" 1`] = `
VFile {
  "contents": "<h1 id=\\"foo\\">Foo<a aria-hidden=\\"true\\" class=\\"anchor-heading icon-link\\" href=\\"#foo\\"></a></h1>
<p>foo test</p>
<hr>
<strong>Children</strong>
<ol>
<li><a href=\\"/notes/foo.ch1\\">Ch1</a></li>
</ol>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`WHEN enableHierarchyDisplay is set to false "HTML: enableHierarchyDisplay: PUBLISHING" 1`] = `
VFile {
  "contents": "<h1 id=\\"foo\\">Foo<a aria-hidden=\\"true\\" class=\\"anchor-heading icon-link\\" href=\\"#foo\\"></a></h1>
<p><a href=\\"/foo.ch1.html\\">Ch1</a></p>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`WHEN enableHierarchyDisplay is set to true "HTML: enableHierarchyDisplay: PUBLISHING" 1`] = `
VFile {
  "contents": "<h1 id=\\"foo\\">Foo<a aria-hidden=\\"true\\" class=\\"anchor-heading icon-link\\" href=\\"#foo\\"></a></h1>
<p><a href=\\"/foo.ch1.html\\">Ch1</a></p>
<hr>
<strong>Children</strong>
<ol>
<li><a href=\\"/notes/foo.ch1\\">Ch1</a></li>
</ol>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`WHEN hierarchyDisplayTitle is set to 'Better Children'  "HTML: hierarchyDisplayTitle: PUBLISHING" 1`] = `
VFile {
  "contents": "<h1 id=\\"foo\\">Foo<a aria-hidden=\\"true\\" class=\\"anchor-heading icon-link\\" href=\\"#foo\\"></a></h1>
<p><a href=\\"/foo.ch1.html\\">Ch1</a></p>
<hr>
<strong>Better Children</strong>
<ol>
<li><a href=\\"/notes/foo.ch1\\">Ch1</a></li>
</ol>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`WHEN note enableChildLinks = false "HTML: noEnableChildLinks: PUBLISHING" 1`] = `
VFile {
  "contents": "<h1 id=\\"foo\\">Foo<a aria-hidden=\\"true\\" class=\\"anchor-heading icon-link\\" href=\\"#foo\\"></a></h1>
<p>foo test</p>",
  "cwd": "<PROJECT_ROOT>",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;
